common/data/src/main/java/org/thingsboard/mqtt/broker/common/data/client/credentials/ScramMqttCredentials.java

```diff
@@ -0,0 +1,125 @@
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Base64;
+import java.util.List;
+
+@Data
+public class ScramMqttCredentials implements HasSinglePubSubAutorizationRules {
+
+    public static final int ITERATIONS_COUNT = 4096;
+
+    private static final int SALT_LENGTH = 16;
+
+    @NoXss
+    private String userName;
+    private byte[] salt;
+    private byte[] serverKey;
+    private byte[] storedKey;
+    private ScramAlgorithm algorithm;
+    private PubSubAuthorizationRules authRules;
+
+    @JsonCreator
+    public ScramMqttCredentials(
+            @JsonProperty("userName") String userName,
+            @JsonProperty("password") String password,
+            @JsonProperty("salt") String salt,
+            @JsonProperty("serverKey") String serverKey,
+            @JsonProperty("storedKey") String storedKey,
+            @JsonProperty("algorithm") ScramAlgorithm algorithm,
+            @JsonProperty("authRules") PubSubAuthorizationRules authRules) throws NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException {
+        this.userName = userName;
+        this.authRules = authRules;
+        this.algorithm = algorithm;
+
+        if (password != null) {
+            this.salt = generateRandomSalt();
+            byte[] saltedPassword = pbkdf2(password, this.salt, algorithm);
+            byte[] clientKey = hmac(saltedPassword, "Client Key", algorithm);
+            this.serverKey = hmac(saltedPassword, "Server Key", algorithm);
+            this.storedKey = hash(clientKey, algorithm);
+        } else {
+            this.salt = Base64.getDecoder().decode(salt);
+            this.serverKey = Base64.getDecoder().decode(serverKey);
+            this.storedKey = Base64.getDecoder().decode(storedKey);
+        }
+    }
+
+    public static ScramMqttCredentials newInstance(
+            String username,
+            String password,
+            ScramAlgorithm algorithm,
+            List<String> authorizationRulePatterns) throws NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException {
+        var authRules = PubSubAuthorizationRules.newInstance(authorizationRulePatterns);
+        return new ScramMqttCredentials(username, password, null, null, null, algorithm, authRules);
+    }
+
+    @JsonGetter("salt")
+    public String getSaltStr() {
+        return Base64.getEncoder().encodeToString(this.salt);
+    }
+
+    @JsonGetter("serverKey")
+    public String getServerKeyStr() {
+        return Base64.getEncoder().encodeToString(this.serverKey);
+    }
+
+    @JsonGetter("storedKey")
+    public String getStoredKeyStr() {
+        return Base64.getEncoder().encodeToString(this.storedKey);
+    }
+
+    private static byte[] generateRandomSalt() {
+        byte[] salt = new byte[SALT_LENGTH];
```

Comment:
Pls reuse StringUtils.RANDOM and StringUtils.generateSafeToken(). You will need to refactor this method a bit for your use case.